packages/app/app/components/RepoSearch.vue

```diff
@@ -1,12 +1,169 @@
 <script lang="ts" setup>
 const search = useSessionStorage("search", "");
+const searchResults = ref<
+  Array<{
+    id: string;
+    name: string;
+    owner: {
+      login: string;
+      avatarUrl: string;
+    };
+    _stats?: {
+      batchCount: number;
+      scannedSoFar: number;
+      matchNumber: number;
+    };
+  }>
+>([]);
+const isLoading = ref(false);
+const searchError = ref<string | null>(null);
+const searchStats = ref<{
+  totalScanned: number;
+  batchCount: number;
+  lastResultIndex: number;
+} | null>(null);
+
+let currentSearchController: AbortController | null = null;
+let currentSearchId = 0;
 
 const throttledSearch = useThrottle(search, 500, true, false);
 
-const { data, status } = useFetch("/api/repo/search", {
-  query: computed(() => ({ text: throttledSearch.value })),
-  immediate: !!throttledSearch.value,
-});
+watch(
+  throttledSearch,
+  async (newValue) => {
+    if (currentSearchController) {
+      // console.log(`[SEARCH-CLIENT] Canceling previous search request`);
+      currentSearchController.abort();
+      currentSearchController = null;
+    }
+
+    if (!newValue) {
+      searchResults.value = [];
+      searchStats.value = null;
+      return;
+    }
+
+    const thisSearchId = ++currentSearchId;
+
+    try {
+      searchResults.value = [];
+      isLoading.value = true;
+      searchError.value = null;
+      searchStats.value = null;
+
+      // console.log(
+      //   `[SEARCH-CLIENT] Starting search #${thisSearchId} for "${newValue}"`,
+      // );
+
+      currentSearchController = new AbortController();
+      const response = await fetch(
+        `/api/repo/search?text=${encodeURIComponent(newValue)}`,
+        {
+          signal: currentSearchController.signal,
+        },
+      );
+
+      if (thisSearchId !== currentSearchId) {
+        // console.log(
+        //   `[SEARCH-CLIENT] Ignoring outdated search #${thisSearchId} results`,
+        // );
+        return;
+      }
+
+      if (!response.ok) {
+        throw new Error(`Search failed: ${response.statusText}`);
+      }
+
+      const reader = response.body?.getReader();
+      if (!reader) {
+        throw new Error("Stream reader not available");
+      }
+
+      // console.log(
+      //   `[SEARCH-CLIENT] Stream reader initialized for "${newValue}" (search #${thisSearchId})`,
+      // );
+      const decoder = new TextDecoder();
+      let buffer = "";
+
+      const seenIds = new Set<string>();
+
+      while (true) {
```

Comment:
instead of a while loop, we can just do `for await (const data of reader)` or something like that.

the code here is too complicated and i believe it can be way shorter and cleaner.